[ttLib/pyftmerge] Handle cmap merge better

yincrash opened the pull request:
Changed the merge algorithm to properly handle cmap subtables
of type 4 and 12 with platform id and encoding ids of 3/1 and 3/10
respectively. All other subtables are not merged and ignored.
The resulting merged cmap table includes a subtable of format 4/3/1
and a format 12 subtable iff there are mappings outside of the BMP.

If one font has two codepoints that point to the same glyph,
and another font has the same code points pointing to two other glyphs,
keep the behavior where the first replacement glyph is stored in 'locl'
and output the third glpyh (or more) to let the user know that they were
dropped, instead of failing to merge the font.

Fixes #444
Fixes #322

yincrash commented:
I was trying to merge Emoji fonts so I could get some color emoji on my Windows 7 machine occasionally when I discovered that `pyftmerge` doesn't output valid Windows font files if there are any code points outside the BMP, so I originally wrote a script to generate a cmap4 table, then wrote this patch to fix it in FontTools. Then got a little carried away to try and fix other cmap merging related errors.

behdad commented on Lib/fontTools/merge_test.py:
```diff
@@ -0,0 +1,84 @@
+class CmapMergeUnitTest(unittest.TestCase):
+	def setUp(self):
+		self.merger = Merger()
+		self.table1 = ttLib.getTableClass('cmap')('cmap')
```
Use ttLib.newTable()

behdad commented on Lib/fontTools/merge.py:
```diff
@@ -359,31 +359,57 @@ def merge(self, m, tables):
 	for table,fontIdx in cmapTables:
-		# TODO handle duplicates.
+		# If there is a format 12 table for the same font, ignore the format 4 table
+		if table.format == 4 and (12, fontIdx) in [(t.format, fontIdx) for t,fontIdx in cmapTables]:
```
How about you move this logic earlier, when we are collecting cmapTables?

behdad commented on Lib/fontTools/merge.py:
```diff
@@ -359,31 +359,57 @@ def merge(self, m, tables):
-	cmapTable.language = 0
-	cmap = cmapTable.cmap
+	# Only merges 4/3/1 and 12/3/10 subtables, ignores all other subtables
+	cmapTables = [(t,fontIdx)
```
Rewrite this as a for loop to accommodate for the logic later on (keep one subtable per font).

behdad commented on Lib/fontTools/merge.py:
```diff
@@ -359,31 +359,57 @@ def merge(self, m, tables):
+				else:
+					# char previously mapped to oldgid but already remapped to a different gid,
+					# save new gid as an alternate
+					# TODO: save the alternates in a 'aalt' table. currently just dropped.
```
Saving in 'aalt' is not a useful option.  Remove mention of that and just print a warning that this uni,gid mapping from font number X is being lost...

We can fix this in most cases if we try harder; but first let's land your patch with existing logic.

yincrash replied:
Removed. The warning output is now like
```
$ pyftmerge NotoNaskhArabic-Regular.ttf NotoSerif-Regular.ttf
WARNING: Dropped mapping from codepoint 0xa0 to glyphId 'uni00A0#1'
WARNING: Dropped mapping from codepoint 0xfeff to glyphId 'uniFEFF#1'
```

Is `aalt` not useful because it's not implemented anywhere, or am I misinterpreting the spec on it?

behdad commented:
Thanks for this!  I left a few minor comments.  Please address those and I'll merge.

yincrash committed:
[ttLib/pyftmerge] Handle cmap merge better

Changed the merge algorithm to properly handle cmap subtables
of type 4 and 12 with platform id and encoding ids of 3/1 and 3/10
respectively. All other subtables are not merged and ignored.
The resulting merged cmap table includes a subtable of format 4/3/1
and a format 12 subtable iff there are mappings outside of the BMP.

If one font has two codepoints that point to the same glyph,
and another font has the same code points pointing to two other glyphs,
keep the behavior where the first replacement glyph is stored in 'locl'
and output the third glpyh (or more) to let the user know that they were
dropped, instead of failing to merge the font.

Fixes #444
Fixes #322

yincrash commented:
Amended my commit with your recommendations :)

behdad commented:
Thanks!
